apps/emqx_ft/test/emqx_ft_storage_exporter_s3_SUITE.erl

```diff
@@ -0,0 +1,132 @@
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+
+-module(emqx_ft_storage_exporter_s3_SUITE).
+
+-compile(export_all).
+-compile(nowarn_export_all).
+
+-include_lib("common_test/include/ct.hrl").
+-include_lib("stdlib/include/assert.hrl").
+
+all() -> emqx_common_test_helpers:all(?MODULE).
+
+init_per_suite(Config0) ->
+    BucketName = create_bucket(),
+    Config1 = [{bucket_name, BucketName} | Config0],
+    ok = emqx_common_test_helpers:start_apps([emqx_ft], set_special_configs(Config1)),
+    Config1.
+end_per_suite(_Config) ->
+    ok = emqx_common_test_helpers:stop_apps([emqx_ft]),
+    ok.
+
+set_special_configs(Config) ->
+    fun
+        (emqx_ft) ->
+            Storage = emqx_ft_test_helpers:local_storage(Config, #{
+                exporter => s3, bucket_name => ?config(bucket_name, Config)
+            }),
+            emqx_ft_test_helpers:load_config(#{<<"enable">> => true, <<"storage">> => Storage});
+        (_) ->
+            ok
+    end.
+
+init_per_testcase(Case, Config) ->
+    ClientId = atom_to_binary(Case),
+    [{clientid, ClientId} | Config].
+end_per_testcase(_Case, _Config) ->
+    ok.
+
+%%--------------------------------------------------------------------
+%% Test Cases
+%%-------------------------------------------------------------------
+
+t_happy_path(Config) ->
+    ClientId = ?config(clientid, Config),
+
+    FileId = <<"🌚"/utf8>>,
+    Name = "cool_name",
+    Data = <<"data"/utf8>>,
+
+    ?assertEqual(
+        ok,
+        emqx_ft_test_helpers:upload_file(ClientId, FileId, Name, Data)
+    ),
+
+    {ok, [#{uri := Uri}]} = emqx_ft_storage:files(),
+
+    ?assertMatch(
+        {ok, {{_StatusLine, 200, "OK"}, _Headers, "data"}},
+        httpc:request(Uri)
+    ),
+
+    Key = binary_to_list(ClientId) ++ "/" ++ binary_to_list(FileId) ++ "/" ++ Name,
+    Meta = erlcloud_s3:get_object_metadata(
+        ?config(bucket_name, Config), Key, emqx_ft_test_helpers:aws_config()
+    ),
+
+    ?assertEqual(
+        ClientId,
+        metadata_field("clientid", Meta)
+    ),
+
+    ?assertEqual(
+        FileId,
+        metadata_field("fileid", Meta)
+    ),
+
+    NameBin = list_to_binary(Name),
+    ?assertMatch(
+        #{
+            <<"name">> := NameBin,
+            <<"size">> := 4
+        },
+        emqx_utils_json:decode(metadata_field("filemeta", Meta), [return_maps])
+    ).
+
+t_upload_error(Config) ->
+    ClientId = ?config(clientid, Config),
+
+    FileId = <<"🌚"/utf8>>,
+    Name = "cool_name",
+    Data = <<"data"/utf8>>,
+
+    OldConfig = emqx_config:get([file_transfer]),
+    NewConfig = emqx_utils_maps:deep_put([storage, exporter, bucket], OldConfig, "invalid-bucket"),
+
+    ok = emqx_ft_conf:on_config_update(OldConfig, NewConfig),
```

Comment:
Q: Why do it like this, instead of `emqx_conf:update([file_transfer, storage, exporter, bucket], <<"invalid-bucket">>, #{})`?

Reply:
✅ 
Didn't want to start `emqx_conf`